proposals/0072-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,247 @@
+---
+simd: '0072'
+title: Feature Gate Threshold Automation
+authors:
+  - Tyera Eulburg
+  - Joe Caulfield
+category: Standard
+type: Core
+status: Draft
+created: 2023-10-11
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+This SIMD outlines a proposal for automating the feature activation process
+based on a stake-weighted support threshold, rather than manual human action.
+
+## Motivation
+
+Feature gates wrap new cluster functionality, and typically change the rules of
+consensus. As such, a feature gate needs to be supported by a strong majority
+of cluster stake when it is activated, or else it risks partitioning the
+network. The current feature-gate system comprises two steps:
+
+1. An individual key-holder queues a feature gate for activation
+2. The runtime automatically activates the feature on the next epoch boundary
+
+The key-holder is the one who (with the help of the solana-cli) assesses the
+amount of stake that recognizes the feature and decides whether it is safe to
+activate. This is obviously brittle and subject to human error.
+
+This SIMD proposes that the runtime itself replaces the human to assess the
+amount of stake that supports a particular queued feature and only
+automatically activates it when a preset threshold is met. 
+
+## New Terminology
+
+- Feature Gate program: The BPF program that will own all feature accounts.
+- Feature-Request account: The PDA under the Feature Gate program used to track
+features requested for activation.
+- Feature-Queue account: The PDA under the Feature Gate program used to track
+features queued for activation that must have support signaled by nodes.
+
+## Detailed Design
+
+Before this SIMD is published, we will deploy a BPF program to manage feature
+activations, which will allow us to revoke pending feature activations
+(previously not possible).
+The feature activation process will otherwise be unchanged, so no SIMD is
+required.
+This program will be upgradeable; the authority and process through which it
+will be upgraded is outside the scope of this SIMD. Since this SIMD requires an
+update to the program, the upgrade details will be determined in parallel.
+
+The initial Feature Gate Program can be found in this pull request to the
+Solana Program Library:
+<https://github.com/solana-labs/solana-program-library/pull/4987>.
+The runtime changes and feature activation that will enable this program can be
+found in this feature tracking issue:
+<https://github.com/solana-labs/solana/issues/33547>.
+
+### Overview
+
+Consider two arbitrary versions of the Solana runtime, where the newer version
+has more features in its feature set.
+
+| v1.14.25  | v1.16.0   |
+| --------- | --------- |
+| Feature A | Feature A |
+|           | Feature B |
+| Feature C | Feature C |
+|           | Feature D |
```

Comment:
What if there are a few "lone dissenters", should they be able to stall the whole network's progress?